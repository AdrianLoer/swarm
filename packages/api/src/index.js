// @flow

import regeneratorRuntime from 'regenerator-runtime';
import hash from 'object-hash';

import Client from 'swarm-client';
import Op, {Frame, ZERO, UUID, FRAME_SEP, js2ron} from 'swarm-ron';
import {ZERO as ZERO_UUID} from 'swarm-ron-uuid';
import {lww, set, ron2js} from 'swarm-rdt';
import type {Atom} from 'swarm-ron';
import type {Options as ClntOpts} from 'swarm-client';

export type Options = ClntOpts & {
  gcPeriod?: number,
};

export type Value = {[string]: Atom | Value} | Value[] | null;

interface Subscription {
  off(): boolean;
  is(hash: string): boolean;
}

export default class API {
  client: Client;
  options: Options;
  subs: Array<Subscription>;
  cache: {[string]: {[string]: Atom}};
  gcInterval: IntervalID;

  constructor(options: Options): API {
    this.client = new Client(options);
    this.options = options;
    this.subs = [];
    this.cache = {};
    if (options.gcPeriod) {
      this.gcInterval = setInterval(this.gc.bind(this), options.gcPeriod);
    }
    return this;
  }

  ensure(): Promise<void> {
    return this.client.ensure();
  }

  uuid(): UUID {
    if (!this.client.clock) throw new Error('have no clock yet, invoke `await <swarm>.ensure()` first');
    return this.client.clock.time();
  }

  async on(id: string | UUID, cbk: (value: Value) => void): Promise<boolean> {
    if (!id || !cbk) return false;
    const h = OnSub.hash(id.toString(), cbk);
    for (const sub of this.subs) {
      if (sub.is(h)) return false;
    }
    const sub = new OnSub(this.client, this.cache, id.toString(), cbk);
    this.subs.push(sub);
    const subscribed = await sub.on();
    return subscribed;
  }

  off(id: string | UUID, cbk: Value => void): boolean {
    if (!id || !cbk) return false;
    let i = -1;
    const h = OnSub.hash(id.toString(), cbk);
    for (const sub of this.subs) {
      i++;
      if (sub.is(h)) {
        this.subs.splice(i, 1);
        return sub.off();
      }
    }
    return false;
  }

  // garbage collection for unused cached data
  gc(): {|deleted: number, existing: number|} {
    const ret = {deleted: 0, existing: 0};
    const lstnrs = {};
    for (const id of Object.keys(this.client.lstn)) lstnrs[id] = true;

    for (const id of Object.keys(this.cache)) {
      if (!lstnrs[id]) {
        delete this.cache[id];
        ret.deleted++;
      } else {
        ret.existing++;
      }
    }

    return ret;
  }

  async set(id: string | UUID, value: {[string]: Atom | void}): Promise<boolean> {
    if (!id) return false;
    const uuid = id instanceof UUID ? id : UUID.fromString(id);
    await this.client.ensure();
    const frame = new Frame();
    let op = new Op(lww.type, uuid, this.uuid(), ZERO_UUID, undefined, FRAME_SEP);
    frame.push(op);

    for (const k of Object.keys(value)) {
      op = op.clone();
      op.location = UUID.fromString(k);
      if (value[k] !== undefined) {
        op.values = js2ron([value[k]]);
        if (!uuid.isLocal() && value[k] instanceof UUID && value[k].isLocal()) {
          return false;
        }
      }
      frame.pushWithTerm(op, ',');
    }

    if (frame.toString()) {
      await this.client.push(frame.toString());
      return true;
    }
    return false;
  }

  async add(id: string | UUID, value: Atom): Promise<boolean> {
    if (!id) return false;
    const uuid = id instanceof UUID ? id : UUID.fromString(id);
    await this.client.ensure();
    const frame = new Frame();
    const time = this.uuid();
    let op = new Op(set.type, uuid, time, ZERO_UUID, undefined, FRAME_SEP);
    frame.push(op);
    op = op.clone();

    if (!uuid.isLocal() && value instanceof UUID && value.isLocal()) {
      return false;
    }

    op.values = js2ron([value]);
    frame.pushWithTerm(op, ',');

    await this.client.push(frame.toString());
    return true;
  }

  async remove(id: string | UUID, value: Atom): Promise<boolean> {
    if (!id) return false;
    const uuid = id instanceof UUID ? id : UUID.fromString(id);
    id = uuid.toString();
    await this.client.ensure();
    const frame = new Frame();
    let deleted = false;
    const ts = this.uuid();
    let op = new Op(set.type, uuid, ts, ZERO_UUID, undefined, FRAME_SEP);
    frame.push(op);

    let state = await this.client.storage.get(id);
    if (!state) {
      state = await this.client.once(`#${id}`);
    }

    if (!state) return false;

    const str = js2ron([value]);
    for (const v of new Frame(state)) {
      if (!v.isRegular()) continue;
      if (v.values === str) {
        deleted = true;
        op = op.clone();
        op.location = v.event;
        op.values = '';
        frame.pushWithTerm(op, ',');
      }
    }

    if (deleted) {
      await this.client.push(frame.toString());
    }
    return deleted;
  }
}

interface IClient {
  on(id: string, cbk: (f: string, s: string) => void): Promise<boolean>;
  off(id: string, cbk: (f: string, s: string) => void): string | void;
}

class OnSub {
  cache: {[string]: {[string]: Atom}};
  client: IClient;
  id: string;
  prev: string;
  cbk: Value => void;
  keys: {[string]: true};
  active: boolean | void;
  hash: string;

  constructor(client: IClient, cache: {[string]: {[string]: Atom}}, id: string, cbk: Value => void): OnSub {
    this.client = client;
    this.cache = cache;
    this.id = id;
    this.cbk = cbk;
    this.hash = OnSub.hash(id, cbk);
    this.prev = new Op(ZERO_UUID, UUID.fromString(id), ZERO_UUID, ZERO_UUID).toString();
    this.keys = {};
    this.keys[this.id] = true;
    // $FlowFixMe
    this._invoke = this._invoke.bind(this);
    return this;
  }

  is(hash: string): boolean {
    return this.hash === hash;
  }

  off(): boolean {
    if (this.active === true) {
      return (this.active = !!this.client.off(this.prev, this._invoke));
    }
    return false;
  }

  async on(): Promise<boolean> {
    if (this.active !== undefined) return false;
    const {frame} = buildTree(this.cache, this.id);
    this.active = await this.client.on((this.prev = frame.toString()), this._invoke);
    return this.active || false;
  }

  _invoke(l: string, s: string): void {
    // TODO handle log and state

    // prevent unauthorized calls
    if (this.active === false) {
      this.client.off('', this._invoke);
      return;
    }
    if (!s) return;
    const v = ron2js(s);
    if (!v) return;

    // $FlowFixMe ?
    this.cache[v.id] = v;
    const {ids, frame, tree} = buildTree(this.cache, this.id);

    if (this.prev !== frame.toString()) {
      this.keys = ids;
      this.client.on(frame.toString(), this._invoke);

      // get the difference and unsubscribe from lost refs
      const off = getOff(ids, this.prev);
      if (off) this.client.off(off, this._invoke);
      this.prev = frame.toString();
    }

    this.cbk(tree);
  }

  static hash(id: string, cbk: Value => void): string {
    return hash({id, cbk});
  }
}

function buildTree(
  cache: *,
  id: string,
  frame: Frame = new Frame(),
  ids: {[string]: true} = {},
): {frame: Frame, tree: Value, ids: {[string]: true}} {
  ids[id] = true;
  frame.push(new Op(ZERO_UUID, UUID.fromString(id), ZERO_UUID, ZERO_UUID));
  let root = cache[id];
  if (!root) return {frame, tree: null, ids};
  // $FlowFixMe
  root = Object.assign(Object.create(Object.getPrototypeOf(root)), root);
  for (const key of Object.keys(root)) {
    const v = root[key];
    if (v instanceof UUID) {
      root[key] = buildTree(cache, v.toString(), frame, ids).tree || Object.freeze(v);
    }
  }
  return {frame, tree: Object.freeze(root), ids};
}

export function getOff(keys: {[string]: true}, ids: string): string {
  const ret = new Frame();
  for (const op of new Frame(ids)) {
    const id = op.object.toString();
    if (!keys[id]) ret.push(op);
  }
  return ret.toString();
}
