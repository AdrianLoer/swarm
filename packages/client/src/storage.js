// @flow

export interface Storage {
  set(key: string, value: string): Promise<void>;
  get(key: string): Promise<string | void>;
  multiGet(keys: string[]): Promise<{[string]: string | void}>;
  remove(key: string): Promise<void>;
  keys(): Promise<Array<string>>;
}

export class InMemory implements Storage {
  storage: {[string]: string};

  constructor(storage: {[string]: string} = {}) {
    this.storage = storage;
  }

  set(key: string, value: string): Promise<void> {
    this.storage[key] = value;
    return Promise.resolve();
  }

  get(key: string): Promise<string | void> {
    return Promise.resolve(this.storage[key]);
  }

  multiGet(keys: string[]): Promise<{[string]: string | void}> {
    const ret = {};
    for (const k of keys) {
      ret[k] = this.storage[k];
    }
    return Promise.resolve(ret);
  }

  remove(key: string): Promise<void> {
    delete this.storage[key];
    return Promise.resolve();
  }

  keys(): Promise<Array<string>> {
    return Promise.resolve(Object.keys(this.storage));
  }
}

export class LocalStorage implements Storage {
  set(key: string, value: string): Promise<void> {
    localStorage.setItem(key, value);
    return Promise.resolve();
  }

  get(key: string): Promise<string | void> {
    const v = localStorage.getItem(key);
    return Promise.resolve(v != null ? v : undefined);
  }

  multiGet(keys: string[]): Promise<{[string]: string | void}> {
    const ret = {};
    for (const k of keys) {
      const item = localStorage.getItem(k);
      ret[k] = item !== null ? item : undefined;
    }
    return Promise.resolve(ret);
  }

  remove(key: string): Promise<void> {
    localStorage.removeItem(key);
    return Promise.resolve();
  }

  keys(): Promise<Array<string>> {
    return Promise.resolve(Object.keys(localStorage));
  }
}
