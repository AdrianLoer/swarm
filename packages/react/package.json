{
  "name": "swarm-react",
  "version": "2.0.0",
  "description": "Swarm to React bindings",
  "author": "Oleg Lebedev <ole6edev@gmail.com> (https://github.com/olebedev)",
  "main": "lib/index.js",
  "repository": {
    "type": "git",
    "url": ""
  },
  "dependencies": {
    "swarm-api": "^2.0.0"
  },
  "devDependencies": {
    "babel-preset-react": "^6.24.1",
    "react-test-renderer": "16.2.0"
  },
  "peerDependencies": {
    "invariant": "^2.2.2",
    "prop-types": "^15.6.0",
    "react": "16.2.0",
    "react-dom": "16.2.0"
  },
  "files": ["*.js", "README.md"],
  "scripts": {
    "build": "yarn run build:clean && yarn run build:lib && yarn run build:flow",
    "build:clean": "../../node_modules/.bin/rimraf lib",
    "build:lib": "../../node_modules/.bin/babel -d lib src --ignore '**/__tests__/**'",
    "build:flow": "../../node_modules/.bin/flow-copy-source -v -i '**/__tests__/**' src lib"
  },
  "keywords": ["swarm", "react", "replicated", "RON", "CRDT"],
  "license": "MIT"
}
