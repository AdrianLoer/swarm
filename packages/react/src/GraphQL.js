// @flow

import * as React from 'react';
import PropTypes from 'prop-types';
import invariant from 'invariant';
import type {DocumentNode} from 'graphql';

import DB from 'swarm-db';
import type {Request, Response} from 'swarm-db';
import type {Value} from 'swarm-api';
import type {Atom} from 'swarm-ron';
import UUID, {ERROR} from 'swarm-ron-uuid';

type args = {[string]: Atom | {[string]: Atom}};

type Props = {
  query: DocumentNode,
  args?: args,
  swarm?: DB,
  mutations?: {[string]: DocumentNode},
  children: ({
    data: Value,
    uuid: () => UUID,
    error?: Error,
    mutations?: {[string]: (args: args) => Promise<Value>},
  }) => React.Node,
};

type State = {
  data: Value,
  error?: Error,
  mutations?: {[string]: (args: args) => Promise<Value>},
};

export default class GraphQL extends React.Component<Props, State> {
  swarm: ?DB;
  _off: void | (() => boolean);

  constructor(props: Props, context: {swarm: ?DB}) {
    super(props, context);
    this.swarm = context.swarm || this.props.swarm;
    invariant(
      this.swarm,
      `Could not find "swarm" in either the context or ` +
        `props of <GraphQL>. ` +
        `Either wrap the root component in a <Provider>, ` +
        `or explicitly pass "swarm" as a prop to <GraphQL>.`,
    );
    invariant(props.query, `Could not find "query" in props of <GraphQL>.`);

    this.state = {
      data: null,
      mutations: this._bindMutations(),
    };

    if (this.swarm) {
      this.swarm
        .ensure()
        .then(this._subscribe.bind(this))
        .catch(error => this.setState({error}));
    }
  }

  componentDidUpdate(prev: Props) {
    if (
      this.props.query !== prev.query ||
      !shallowEqual(this.props.args, prev.args) ||
      !shallowEqual(this.props.mutations, prev.mutations)
    ) {
      this._unsubscribe();
      this.setState(
        {
          data: null,
          error: undefined,
          mutations: this._bindMutations(),
        },
        this._subscribe.bind(this),
      );
    }
  }

  componentWillUnmount() {
    this._unsubscribe();
  }

  async _subscribe(): Promise<void> {
    const {props: {query, args}, swarm} = this;
    if (!swarm || !swarm.execute) return;

    const sub = await swarm.execute({gql: query, args}, (r: Response) => {
      this.setState({data: r.data, error: r.error});
    });

    if (this._off) this._off();
    this._off = sub.off;
  }

  _unsubscribe() {
    this._off && this._off();
  }

  _bindMutations(): {[string]: (args: args) => Promise<Value>} | void {
    const {props: {mutations}, swarm} = this;
    if (mutations && swarm) {
      const ret = {};
      for (const key of Object.keys(mutations)) {
        ret[key] = async (args: args): Promise<Value> => {
          return new Promise((resolve, reject) => {
            swarm.execute({gql: mutations[key], args: args}, (r: Response) => {
              r.error ? reject(r.error) : resolve(r.data);
            });
          });
        };
      }
      return ret;
    }
  }

  render() {
    return (
      this.props.children &&
      this.props.children.call(null, {
        data: this.state.data,
        uuid: this.swarm ? this.swarm.uuid : () => ERROR,
        error: this.state.error,
        mutations: this.state.mutations,
      })
    );
  }
}

GraphQL.contextTypes = {
  swarm: PropTypes.shape({}).isRequired,
};

function shallowEqual(a?: {}, b?: {}): boolean {
  if (a === b) return true;
  if (a == null || b == null) return false;
  var aKeys = Object.keys(a);
  if (Object.keys(b).length !== aKeys.length) return false;
  for (const key of aKeys) {
    if (a[key] !== b[key]) return false;
  }
  return true;
}
