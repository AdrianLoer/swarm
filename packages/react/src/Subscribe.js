// @flow

import * as React from 'react';
import PropTypes from 'prop-types';
import invariant from 'invariant';

import API from 'swarm-api';
import type {Value} from 'swarm-api';

type Props = {
  to: Array<string>,
  swarm?: API,
  children?: ({
    unsubscribe: () => void,
    initialized: boolean,
    swarm: ?API,
    data: {[string]: Value},
  }) => React.Node,
};

type State = {
  initialized: boolean,
  data: {[string]: Value},
};

export default class Subscribe extends React.Component<Props, State> {
  swarm: ?API;
  cbks: {[string]: (Value) => void};

  constructor(props: Props, context: {swarm: ?API}) {
    super(props, context);
    this.swarm = context.swarm || this.props.swarm;
    invariant(
      this.swarm,
      `Could not find "swarm" in either the context or ` +
        `props of <Subscribe>. ` +
        `Either wrap the root component in a <Provider>, ` +
        `or explicitly pass "swarm" as a prop to <Subscribe>.`,
    );
    this.cbks = {};
    this.state = {
      initialized: !!this.swarm && !!this.swarm.client.clock,
      data: {},
    };

    if (!this.state.initialized && !!this.swarm) {
      this.swarm
        .ensure()
        .then(() => {
          this._subscribe();
          this.setState({initialized: true});
        })
        .catch(e => console.error(e));
    } else {
      this._subscribe();
    }

    // $FlowFixMe
    this._unsubscribe = this._unsubscribe.bind(this);
  }

  _subscribe(): void {
    if (!this.swarm) return;
    for (const k of this.props.to) {
      this.cbks[k] = (v: Value) => {
        const state = {
          ...this.state,
          data: {
            ...this.state.data,
            [k]: v || null,
          },
        };
        this.setState(state);
      };

      this.swarm && this.swarm.on(k, this.cbks[k]).catch(e => console.error(e));
    }
  }

  componentDidUpdate(prev: Props) {
    if (!shallowEqual(this.props.to, prev.to)) {
      this._unsubscribe();
      this._subscribe();
    }
  }

  componentWillUnmount() {
    this._unsubscribe();
  }

  _unsubscribe(): void {
    for (const k of Object.keys(this.cbks)) {
      if (this.swarm) this.swarm.off(k, this.cbks[k]);
    }
    this.cbks = {};
  }

  _filterData(): {[string]: Value} {
    const ret = {};
    for (const k of this.props.to) {
      ret[k] = this.state.data[k] || null;
    }
    return ret;
  }

  render() {
    return (
      this.props.children &&
      this.props.children.call(null, {
        unsubscribe: this._unsubscribe,
        initialized: this.state.initialized,
        swarm: this.swarm,
        data: this._filterData(),
      })
    );
  }
}

Subscribe.contextTypes = {
  swarm: PropTypes.shape({}).isRequired,
};

function shallowEqual(a: string[], b: string[]): boolean {
  if (!Array.isArray(a) || !Array.isArray(b)) return false;
  if (a === b) return true;
  if (a.length !== b.length) return false;
  for (const item of a) {
    if (b.indexOf(item) === -1) return false;
  }
  return true;
}
