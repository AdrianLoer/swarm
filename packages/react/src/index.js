// @flow

export {default as Provider} from './Provider';
export {default as Subscribe} from './Subscribe';
