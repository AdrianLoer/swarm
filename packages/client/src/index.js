/* @flow */
'use strict';

import RWS from 'reconnectable-websocket';

import Op, {Frame, Cursor, UUID, QUERY_SEP, FRAME_SEP, mapUUIDs, js2ron} from 'swarm-ron';
import type {Clock} from 'swarm-clock';
import {Logical} from 'swarm-clock';
import {ZERO, NEVER} from 'swarm-ron-uuid';
import {reduce} from 'swarm-rdt';
import type {Storage} from './storage';
import {InMemory} from './storage';

export interface Connection {
  onmessage: (ev: MessageEvent) => any;
  onopen: (ev: Event) => any;
  send(data: string): void;
  readyState: number;
}

export type Meta = {
  name: string,
  clockMode?: 'Logical' | 'Epoch' | 'Calendar',
  clockLen?: number,
  forkMode?: string,
  peerIdBits?: number,
  horizont?: number,
  credentials?: {},
  seen?: string,
};

export type Options = {
  id: string,
  storage: Storage,
  upstream?: string | Connection,
  hsTimeout?: number,
  db?: Meta,
};

const defaultMeta: Meta = {
  name: 'default',
  clockLen: 5,
  forkMode: '// FIXME', // keep it sync with server
  peerIdBits: 30,
  horizont: 604800, // one week in seconds
};

/**
 * A simple client, keeps data in memory.
 * Consumes updates from the server, feeds resulting RON states
 * back to the listeners.
 */
export default class Client {
  log: Frame; // FIXME ??
  updates: Array<string>;

  clock: ?Clock;
  lstn: {[key: string]: (frame: string, state: string) => void}; // ?
  upstream: Connection;
  storage: Storage;
  queue: Array<[() => void, (err: Error) => void]> | void;
  db: Meta;
  id: string;
  options: {
    hsTimeout: number,
  };

  constructor(options: Options) {
    this.id = options.id;
    this.db = {
      ...defaultMeta,
      ...options.db,
    };
    this.storage = options.storage;
    this.lstn = {};
    this.log = new Frame();
    this.queue = [];
    this.updates = [];
    this.options = {
      hsTimeout: options.hsTimeout || 3e5 /* 5min */,
    };
    this.init(options);
  }

  async init(options: Options) {
    try {
      const meta = await this.storage.get('__meta__');
      this.db = ({
        ...this.db,
        ...JSON.parse(meta || '{}'),
      }: Meta);

      if (this.db.clockMode) {
        switch (this.db.clockMode) {
          case 'Logical':
            this.clock = new Logical(this.id);
            await this.storage.set('__meta__', JSON.stringify(this.db));
            break;
          default:
            throw new Error(`TODO: Clock mode '${this.db.clockMode}' is not supported yet`);
        }
      }

      if (typeof options.upstream === 'string') {
        this.upstream = RWS(options.upstream, undefined, {reconnectOnError: true});
      } else if (options.upstream) {
        this.upstream = options.upstream;
      } else if (!options.upstream && this.db.clockMode) {
        this.upstream = new DevNull();
      } else if (!options.upstream && !this.db.clockMode) {
        throw new Error('neither connection options nor clock options found');
      }

      await new Promise(r => {
        this.upstream.onopen = r;
      });
      await this.handshake();
      this.release(null);
    } catch (e) {
      this.release(e);
    }
  }

  async handshake(): Promise<void> {
    if (this.upstream instanceof DevNull) return;

    const hs = new Promise((resolve, reject) => {
      setTimeout(reject, this.options.hsTimeout);
      this.upstream.onmessage = (me: MessageEvent) => resolve(((me.data: any): string));
    });

    const hello = new Frame();
    hello.push(
      new Op(
        new UUID('db', '0', '$'),
        new UUID(this.db.name, '0', '$'),
        new UUID(this.clock ? this.clock.last().value : '0', this.id, '+'),
        ZERO,
        QUERY_SEP,
      ),
    );
    hello.push(new Op(ZERO, ZERO, ZERO, ZERO, FRAME_SEP));

    const creds = this.db.credentials || {};
    for (const cred of Object.keys(creds)) {
      hello.push(new Op(ZERO, ZERO, ZERO, UUID.fromString(cred), js2ron([creds[cred]])));
    }
    this.upstream.send(hello.toString());

    const resp = await hs;

    const dbOpts: Meta = {
      clockMode: 'Logical',
      ...this.db,
      credentials: {...this.db.credentials},
    };

    let seen: UUID = ZERO;
    for (const op of new Frame(resp)) {
      if (op.uuid(3).gt(seen)) seen = op.uuid(3);
      const val = op.value(0);
      if (val) {
        let key = op.uuid(3).toString();
        key = key[0].toLowerCase() + key.slice(1);
        dbOpts[key] = val instanceof UUID ? val.value : val;
      }
    }

    if (this.clock) {
      if (this.db.clockMode !== dbOpts.clockMode)
        throw new Error(
          `Different clock mode: '${this.db.clockMode || 'undefined'}' !== '${dbOpts.clockMode || 'undefined'}'`,
        );
    } else {
      switch (dbOpts.clockMode) {
        case 'Logical':
          this.clock = new Logical(this.id);
        default:
          throw new Error(`TODO: Clock mode '${dbOpts.clockMode || 'undefined'}' is not supported yet`);
      }
    }

    this.clock.see(seen);

    this.db = {...this.db, ...dbOpts};
    await this.storage.set('__meta__', JSON.stringify(this.db));
    this.upstream.onmessage = (me: MessageEvent) => this.onMessage(((me.data: any): string)).catch(panic);
    this.upstream.onopen = () => this.handshake().catch(panic);

    // resend all the frames
    for (const frame of this.updates) {
      this.upstream.send(frame);
    }

    // resubscribe
    let query: string = '';
    for (const key of Object.keys(this.lstn)) {
      query += key;
    }
    await this.on(query);
  }

  /**
   * Ensure returns Promise which will be resolved after connection
   * installed or rejected if an error occurred.
   */
  async ensure(): Promise<void> {
    const {queue} = this;
    if (queue === undefined) {
      return Promise.resolve();
    } else {
      return new Promise((release, reject) => {
        queue.push([release, reject]);
      });
    }
  }

  release(err: Error | null) {
    if (!this.queue) return;
    for (const cbk of this.queue) {
      if (err) {
        cbk[1](err);
      } else {
        cbk[0]();
      }
    }
    delete this.queue;
  }

  async onMessage(message: string): Promise<void> {
    if (!this.clock) throw new Error('Have no clock');
    for (const op of new Frame(message)) {
      if (op.event.origin === this.clock.origin()) {
        for (let i = this.updates.length - 1; i >= 0; i--) {
          const update = Op.fromString(this.updates[i]);
          if (update && op.event.lt(update.event)) continue;
          this.updates = this.updates.slice(i);
          i = -1;
        }
      }
    }
    await this.update(message);
  }

  /**
   * On installs subscriptions.
   */
  async on(query: string, callback: ?(frame: string, state: string) => void): Promise<void> {
    const fwd = new Frame();
    for (const op of new Frame(query)) {
      const key = op.key();
      let base = ZERO;
      const stored = await this.storage.get(key);
      if (stored) {
        for (const op of new Frame(stored)) {
          base = op.event;
          break;
        }
      }
      fwd.push(new Op(op.type, op.object, base, ZERO, QUERY_SEP));
      if (callback) {
        if (this.lstn[key]) throw new Error('TODO: many listeners per obj');
        this.lstn[key] = callback;
      }
    }

    this.upstream.send(fwd.toString());
    if (callback) await this.update(fwd.toString(), true);
  }

  /**
   * Off removes subscriptions.
   */
  off(query: string, callback: ?(frame: string, state: string) => void) {
    const fwd = new Frame();
    for (const op of new Frame(query)) {
      delete this.lstn[op.key()];
      fwd.push(new Op(op.type, op.object, NEVER, ZERO));
    }
    this.upstream.send(fwd.toString());
  }

  /**
   * Push sends updates to remote and local storages.
   * Waits for connection installed. Thus, the client works in
   * read-only mode until installed connection.
   */
  async push(rawFrame: string) {
    await this.ensure();
    const stamps: {[string]: UUID} = {};

    // replace
    const frame = mapUUIDs(rawFrame, (uuid, position, index, op): UUID => {
      if ([1, 2].indexOf(position) === -1) return uuid;
      if (!uuid.isName() || !uuid.isZero()) return uuid;

      const key = `${index}/${op.key()}`;
      if (stamps[key]) return stamps[key];
      // $FlowFixMe this.clock
      return (stamps[key] = this.clock.time());
    });

    this.updates.push(frame);

    // save
    const op = Op.fromString(frame);
    if (op) this.log.push(op); // TODO ?
    this.upstream.send(frame);
    await this.update(frame);
  }

  /**
   * Update updates local states and notifies listeners.
   */
  async update(frame: string, skipMerge: ?true): Promise<void> {
    for (const op of new Frame(frame)) {
      const key = op.key();
      let state = await this.storage.get(key);
      if (!skipMerge) {
        state = state ? reduce(state.toString(), frame) : frame;
        await this.storage.set(key, state);
      }
      const l = this.lstn[key];
      if (l) l(frame, state.toString());
      break; // read only first operation of given frame
    }
  }
}

// DevNull connection is used for permanent offline-mode
class DevNull implements Connection {
  onmessage: (ev: MessageEvent) => any;
  onopen: (ev: Event) => any;
  readyState: number;
  constructor() {
    this.readyState = 3;
  }
  send(data: string): void {}
}

function panic(err: any) {
  throw err;
}
