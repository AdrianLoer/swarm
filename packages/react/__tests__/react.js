// @flow

import * as React from 'react';
import renderer from 'react-test-renderer';

import API from 'swarm-api';
import {Provider, Subscribe} from '../src';
import {Connection} from '../../__tests__/fixtures';
import {InMemory} from '../../client/src/storage';

const Basic = ({data, initialized}) => <div initialized={initialized}>{JSON.stringify(data)}</div>;

test('React, basic', async () => {
  const upstream = new Connection('011-react.ron');

  // to keep the connection closed after initialization explicitly
  await new Promise(r => setTimeout(r, 10));

  const storage = new InMemory();
  const api = new API({
    storage,
    upstream,
    db: {
      id: 'user',
      name: 'test',
      auth: 'JwT.t0k.en',
      clockMode: 'Logical',
    },
  });

  let object = {};
  api.on('object', o => {
    object = o;
  });

  // count calls, to unsubscribe after step #3
  let c = -1;

  const component = renderer.create(
    <Provider swarm={api}>
      <Subscribe to={['object']}>
        {props => {
          expect(props.swarm).toBe(api);
          c++;
          if (c === 2) props.unsubscribe();
          // try to return malformed markup, but won't happened actually
          // b/c the state won't changed and next call won't happened
          // b/c of props.unsubscribe call
          if (c > 2) return <span>¯\_(ツ)_/¯</span>;
          return <Basic {...props} />;
        }}
      </Subscribe>
    </Provider>,
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  await new Promise(r => setTimeout(r, 10));

  // start to communicate
  upstream.onopen(new Event(''));

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  await api.ensure();

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // b/c of async nature of mock connection
  await new Promise(r => setTimeout(r, 500));

  expect(object).toEqual({test: 5});

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  await api.set('object', {some: 'value'});
  expect(object).toEqual({test: 5, some: 'value'});

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  const dump = upstream.dump();
  expect(dump.session).toEqual(dump.fixtures);
});
