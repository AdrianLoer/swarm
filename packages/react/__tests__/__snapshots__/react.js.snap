// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React, basic 1`] = `
<div
  initialized={false}
>
  {"object":null}
</div>
`;

exports[`React, basic 2`] = `
<div
  initialized={false}
>
  {"object":null}
</div>
`;

exports[`React, basic 3`] = `
<div
  initialized={true}
>
  {"object":null}
</div>
`;

exports[`React, basic 4`] = `
<div
  initialized={true}
>
  {"object":{"test":5}}
</div>
`;

exports[`React, basic 5`] = `
<div
  initialized={true}
>
  {"object":{"test":5}}
</div>
`;
